### PR TITLE
Iterative instead of recursive implementation, @inline(__always) a few more functions

### PR DESCRIPTION
**Pull request by hassila**
Experiement with iterative implementations for core functions plus some extra inlining, seems to give a fairly significant 15-20% improvement.

### Checklist
- [ ] I've read the [Contribution Guidelines](/README.md#contributing-to-swift-collections)
- [x ] My contributions are licensed under the [Swift license](/LICENSE.txt).
- [ x] I've followed the coding style of the rest of the project.
- [ ] I've added tests covering all new code paths my change adds to the project (if appropriate).
- [ ] I've added benchmarks covering new functionality (if appropriate).
- [x ] I've verified that my change does not break any existing tests or introduce unexplained benchmark regressions.
- [ ] I've updated the documentation if necessary.

Previous recursive implementation benchmark:

priorityqueue-recursive

Iterative implementation results:
priorityqueue-iterative

### CODE REVIEW

**Comment by hassila** [comment]
[Benchmark differentials.pdf](https://github.com/AquaGeek/swift-collections/files/6703270/Benchmark.differentials.pdf)

**Comment by hassila** [comment]
[Benchmark differentials.pdf](https://github.com/AquaGeek/swift-collections/files/6703299/Benchmark.differentials.pdf)

File: Sources/PriorityQueueModule/PriorityQueue.swift [hunk 1]
```diff
@@ -173,23 +173,33 @@ public struct PriorityQueue<Element: Comparable> {
...
   internal mutating func _bubbleUpMin(startingAt index: Int) {
-    guard let grandparentIdx = _grandparentIndex(of: index) else { return }
-
-    if storage[index] < storage[grandparentIdx] {
-      _swapAt(index, grandparentIdx)
-      _bubbleUpMin(startingAt: grandparentIdx)
+    var index = index
+      
+    while true {
+      guard let grandparentIdx = _grandparentIndex(of: index) else { return }
```
Comment (AquaGeek) [inline Sources/PriorityQueueModule/PriorityQueue.swift]: Can this be restructured to be the `while` condition?

```swift
while let grandparentIdx = _grandparentIndex(of: index) {
```

File: Sources/PriorityQueueModule/PriorityQueue.swift [hunk 2]
```diff
@@ -173,23 +173,33 @@ public struct PriorityQueue<Element: Comparable> {
...
+
+      if storage[index] < storage[grandparentIdx] {
+          _swapAt(index, grandparentIdx)
+          index = grandparentIdx
+      } else { return }
```
Comment (AquaGeek) [inline Sources/PriorityQueueModule/PriorityQueue.swift]: Should this comparison also be part of the `while` condition?

```swift
while let grandparentIdx = _grandparentIndex(of: index),
      storage[index] < storage[grandparentIdx] {
```

File: Sources/PriorityQueueModule/PriorityQueue.swift [hunk 3]
```diff
@@ -173,23 +173,33 @@ public struct PriorityQueue<Element: Comparable> {
...
-      _bubbleUpMax(startingAt: grandparentIdx)
+    var index = index
+      
+    while true {
+      guard let grandparentIdx = _grandparentIndex(of: index) else { return }
+
+      if storage[index] > storage[grandparentIdx] {
+        _swapAt(index, grandparentIdx)
+        index = grandparentIdx
+      } else { return }
```
Comment (AquaGeek) [inline Sources/PriorityQueueModule/PriorityQueue.swift]: Same note here about consolidating this to the `while` condition.

File: Sources/PriorityQueueModule/PriorityQueue.swift [hunk 4]
```diff
@@ -214,6 +224,7 @@ public struct PriorityQueue<Element: Comparable> {
 
   // MARK: -
 
+  @inline(__always)
```
Comment (AquaGeek) [inline Sources/PriorityQueueModule/PriorityQueue.swift:222]: Should we add this attribute to `_bubbleUp(startingAt:)` as well?

File: Sources/PriorityQueueModule/PriorityQueue.swift [hunk 5]
```diff
@@ -225,70 +236,76 @@ public struct PriorityQueue<Element: Comparable> {
...
-        _trickleDownMin(startingAt: smallestDescendantIdx)
+      if isChild {
+        return
+      }
+        
+      // Smallest is a grandchild
+      let parentIdx = _parentIndex(of: smallestDescendantIdx)!
+      if storage[smallestDescendantIdx] > storage[parentIdx] {
+        _swapAt(smallestDescendantIdx, parentIdx)
       }
```
Comment (AquaGeek) [inline Sources/PriorityQueueModule/PriorityQueue.swift]: Doesn't this still need to be wrapped in the `if storage[smallestDescendantIdx] < storage[index]` check? We're potentially doing unnecessary work here, aren't we?

File: Sources/PriorityQueueModule/PriorityQueue.swift (hunk 1 above)
Comment (hassila) [reply Sources/PriorityQueueModule/PriorityQueue.swift]: Yep, done in ca63256, thanks.

File: Sources/PriorityQueueModule/PriorityQueue.swift (hunk 2 above)
Comment (hassila) [reply Sources/PriorityQueueModule/PriorityQueue.swift]: Yep, done in ca63256, thanks, nicer!

File: Sources/PriorityQueueModule/PriorityQueue.swift (hunk 3 above)
Comment (hassila) [reply Sources/PriorityQueueModule/PriorityQueue.swift]: Done.

File: Sources/PriorityQueueModule/PriorityQueue.swift (hunk 4 above)
Comment (hassila) [reply Sources/PriorityQueueModule/PriorityQueue.swift:222]: I added it, saw no major difference - but it's part of ca63256 too.

File: Sources/PriorityQueueModule/PriorityQueue.swift (hunk 5 above)
Comment (hassila) [reply Sources/PriorityQueueModule/PriorityQueue.swift]: Yeah, good catch, mea culpa. I've moved the finishing brace and added a return instead, please have another look now.

**Comment by hassila** [comment]
Thanks, addressed all PR feedback in ca63256 now.

**Comment by hassila** [comment]
The incorrectly placed brace made the improvement even nicer, thanks!

```
swift-collections-benchmark results compare bench-recursive bench-new --output diff.html
Tasks with difference scores larger than 1.05:
  Score   Sum     Improvements Regressions  Name
  1.356   1.356   1.356(#75)   0.9994(#1)   PriorityQueue<Int> insert (*)
  1.289   1.289   1.289(#76)   1.000(#0)    PriorityQueue<Int> init from range (*)
  1.204   1.204   1.205(#73)   0.9991(#3)   PriorityQueue<Int> popMin (*)
  1.200   1.200   1.200(#74)   0.9997(#2)   PriorityQueue<Int> popMax (*)
```

[Benchmark differentials.pdf](https://github.com/AquaGeek/swift-collections/files/6704117/Benchmark.differentials.pdf)

**Pull request merged**